raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -584,211 +721,244 @@ private void updateVoterConnections(FindQuorumResponseData response) {
-        if (epoch < quorum.epoch()) {
-            return true;
+    private void maybeBecomeFollower(
+        OptionalInt leaderId,
+        int epoch
+    ) throws IOException {
+        if (!hasConsistentLeader(epoch, leaderId)) {
+            throw new IllegalStateException("Received response with leader " + leaderId +
+                " and epoch " + epoch + " which is inconsistent with current leader " +
+                quorum.leaderId() + " and epoch " + quorum.epoch());
         } else if (epoch > quorum.epoch()) {
             // For any request type, if the response indicates a higher epoch, we bump our local
             // epoch and become a follower. Responses only include elected leaders.
             GracefulShutdown gracefulShutdown = shutdown.get();
             if (gracefulShutdown != null) {
                 gracefulShutdown.onEpochUpdate(epoch);
             } else {
-                becomeFollower(epoch, leaderId);
+                becomeFollower(leaderId, epoch);
             }
-
-            return true;
-        } else if (hasInconsistentLeader(epoch, leaderId)) {
-            throw new IllegalStateException("Received response with leader " + leaderId +
-                    " which is inconsistent with current leader " + quorum.leaderId());
+        } else if (leaderId.isPresent() && !quorum.hasLeader()) {
+            // The response indicates the leader of the current epoch, which is currently unknown
+            becomeFetchingFollower(leaderId.getAsInt(), epoch);
         }
+    }
+
+    private boolean handleUnexpectedError(Errors error, RaftResponse.Inbound response) {
+        logger.error("Unexpected error {} in {} response: {}",
+            error, response.data.apiKey(), response);
         return false;
     }
 
-    private void handleResponse(RaftResponse.Inbound response) throws IOException {
+    private void handleResponse(RaftResponse.Inbound response, long currentTimeMs) throws IOException {
         // The response epoch matches the local epoch, so we can handle the response
-        ApiMessage responseData = response.data();
-        if (responseData instanceof FetchQuorumRecordsResponseData) {
-            handleFetchQuorumRecordsResponse((FetchQuorumRecordsResponseData) responseData);
-        } else if (responseData instanceof VoteResponseData) {
-            handleVoteResponse(response.sourceId(), (VoteResponseData) responseData);
-        } else if (responseData instanceof BeginQuorumEpochResponseData) {
-            handleBeginQuorumEpochResponse(response.sourceId());
-        } else if (responseData instanceof EndQuorumEpochResponseData) {
-            // Nothing to do for now
-        } else if (responseData instanceof FindQuorumResponseData) {
-            handleFindQuorumResponse((FindQuorumResponseData) responseData);
+        ApiKeys apiKey = ApiKeys.forId(response.data.apiKey());
```

Comment:
nit: add log.trace recording the recipient of the response?

Reply:
There is a trace message under `sendOutboundMessage`.